### PR TITLE
[webui][api] fix the color of the build flag preview when changing a flag

### PR DESCRIPTION
**Pull request by mmohring**
After correcting the flag computation in the WebUI, now also fix the colors of the previewed flag change in the projects or packages repository flag view. Taking a default should be grey, and a flag change causing to add or delete or toggle a flag should be blue.

### CODE REVIEW

**Comment by mmohring** [comment]
Please dont merge yet.

**Comment by mmohring** [comment]
@bgeuken @hennevogel @adrianschroeter  This patch is now ready to merge. I have added lots of flags inheritance tests to check also the corner cases. Both the default and the effective flags are now correctly computed and used in the code, e.g. in the API flags class and in the WebUI project => repository => flags section.

**Comment by mmohring** [comment]
Ping ?

File: src/api/app/models/flag.rb [hunk 1]
```diff
@@ -70,6 +70,40 @@ def default_status
...
+    if main_object.kind_of? Package
+      # do the same_flag check first to see if all_flag or same_flag had been set on package level, they *both* overwrite the project level
+      same_flag = main_object.project.flags.where("flag = ? AND repo = ? AND architecture_id = ?", flag, repo, architecture_id).first unless
+        all_flag || same_flag || repo_flag || arch_flag
+      repo_flag = main_object.project.flags.where("flag = ? AND repo = ? AND architecture_id IS NULL", flag, repo).first unless
+        all_flag || repo_flag || arch_flag
+      arch_flag = main_object.project.flags.where("flag = ? AND repo IS NULL AND architecture_id = ?", flag, architecture_id).first unless
+        all_flag || arch_flag
+      all_flag = main_object.project.flags.where("flag = ? AND repo IS NULL AND architecture_id IS NULL", flag).first unless all_flag
+    end
```
Comment (bgeuken) [inline src/api/app/models/flag.rb]: This part looks like a copy & paste from default_status. If so, please move that part into a method and call it in effective_status and default_status

File: src/api/app/models/flag.rb [hunk 2]
```diff
@@ -70,6 +70,40 @@ def default_status
...
+    if arch_flag
+      return all_flag.status if all_flag
+    end
+    if repo_flag
+      return all_flag.status if all_flag
+    end
+    if main_object.kind_of? Package
+      return all_flag.status if all_flag
+    end
```
Comment (bgeuken) [inline src/api/app/models/flag.rb]: The last three ifs can be rewritten as
```ruby
if all_flag
  if arch_flag || repo_flag || main_object.kind_of?(Package)
    return all_flag.status
  end
end
```

File: src/api/app/models/flag.rb [hunk 3]
```diff
@@ -66,7 +66,7 @@ def default_status
       arch_flag = main_object.project.flags.where("flag = ? AND repo IS NULL AND architecture_id = ?", flag, architecture_id).first unless arch_flag
       all_flag = main_object.project.flags.where("flag = ? AND repo IS NULL AND architecture_id IS NULL", flag).first unless all_flag
     end
-    
+
```
Comment (bgeuken) [inline src/api/app/models/flag.rb]: Looks like you introduced those whitespaces in 6ee6edf4fe03b27c7f2942 . Please squash this commit into the one that introduced this change

File: src/api/app/models/flag.rb [hunk 4]
```diff
@@ -79,6 +79,7 @@ def default_status
       return all_flag.status if all_flag
     end
     if main_object.kind_of? Package
+      all_flag = main_object.project.flags.where("flag = ? AND repo IS NULL AND architecture_id IS NULL", flag).first
```
Comment (bgeuken) [inline src/api/app/models/flag.rb]: Please add a description to the commit message explaining why this change was done

File: src/api/app/models/flag.rb (hunk 2 above)
Comment (mmohring) [inline src/api/app/models/flag.rb]: OK. Will rethink.

**Comment by mmohring** [comment]
I will try to unify the code of {effective|default}_status into one routine with a parameter.

**Comment by mmohring** [comment]
Hi, from my side this is now ready to pull. Test suite worked without error, and I have also tested it manually in the WebUI. Works fine now.

File: src/api/test/unit/code_quality_test.rb [hunk 5]
```diff
@@ -87,7 +87,7 @@ def setup
       'PersonController#internal_register'                                      => 112.01,
       'Package#find_changed_issues'                                             => 93.74,
       'Package#close_requests'                                                  => 84.82,
-      'Flag#default_status'                                                     => 90.56,
+      'Flag#compute_status'                                                     => 145.55,
```
Comment (hennevogel) [inline src/api/test/unit/code_quality_test.rb]: Phew...

Comment (mmohring) [inline src/api/test/unit/code_quality_test.rb]: Means what :)

Comment (hennevogel) [inline src/api/test/unit/code_quality_test.rb]: I know it's complicated. I also know the reasons. This was just an expression about my feelings for a complexity of 150 :-)

**Comment by hennevogel** [comment]
Okay let's go

**Comment by bgeuken** [comment]
Thanks for diving into this :+1:

File: src/api/test/unit/code_quality_test.rb (hunk 5 above)
Comment (mmohring) [inline src/api/test/unit/code_quality_test.rb]: OK. I thought you found something else. I thought long about how to simplify the many if's and unless expressions. I used the tricks from the former code. If I find a quiet minute maybe I check if the monster can be simplified somehow :)
